metaconfig-typesafe-config/src/main/scala/metaconfig/typesafeconfig/TypesafeConfig2Class.scala

```diff
@@ -53,36 +53,73 @@ object TypesafeConfig2Class {
               )
           }
       }
-      getPositionOpt(value.origin, input) match {
+      getPositionOpt(value.origin, cache) match {
         case Some(pos) => conf.withPos(pos)
         case None => conf
       }
     }
     try {
-      Configured.Ok(loop(config().resolve().root))
+      val resolved = config().resolve().root()
+      cache += OriginId(resolved.origin) -> input
+      Configured.Ok(loop(resolved))
     } catch {
       case e: ConfigException.Parse =>
         Configured.NotOk(
-          ConfError.parseError(getPosition(e.origin, input), e.getMessage)
+          ConfError.parseError(getPosition(e.origin, cache), e.getMessage)
         )
     }
   }
 
   private def getPosition(
       originOrNull: ConfigOrigin,
-      input: Input
+      cache: mutable.Map[OriginId, Input]
   ): Position =
-    getPositionOpt(originOrNull, input).getOrElse(Position.None)
+    getPositionOpt(originOrNull, cache).getOrElse(Position.None)
 
   private def getPositionOpt(
       originOrNull: ConfigOrigin,
-      input: Input
+      cache: mutable.Map[OriginId, Input]
   ): Option[Position] =
     for {
       origin <- Option(originOrNull)
-      linePlus1 <- Option(origin.lineNumber)
-      line = linePlus1 - 1
-      start = input.lineToOffset(line)
+      input <- cache.updateWith(OriginId(origin)) { maybeCurrent =>
+        maybeCurrent
+          .orElse(
+            Option(origin.url)
+              .map(url => Input.File(new java.io.File(url.toURI)))
```

Comment:
It’s best to use Input.VirtualFile to avoid hashCode/equals issues when comparing inputs. Input.File should never be used because it has the same hashCode even if the underlying file contents change.

Reply:
Good point! I guess in that case it's not a big deal as the `Input` instances are not used as keys and don't escape private methods? I don't see any factory method for creating a Input.VirtualFile out of a Input.File so I wonder if the extra steps are really worth it here.

Reply:
Sounds good, then it should be fine to use `Input.File` here. I think it would make sense to deprecate `Input.File` at some point since it's easy to screw up these things resulting in surprising behavior